CMakeLists.txt

```diff
@@ -18,39 +18,41 @@ endif()
-mxnet_option(USE_F16C             "Build with x86 F16C instruction support" ON) # autodetects support if ON
-mxnet_option(USE_LAPACK           "Build with lapack support" ON)
-mxnet_option(USE_MKL_IF_AVAILABLE "Use MKL if found" ON)
-mxnet_option(USE_MKLDNN           "Build with MKL-DNN support" ON IF USE_MKL_IF_AVAILABLE AND (NOT APPLE) AND (NOT MSVC) AND (CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "x86_64") AND (NOT CMAKE_CROSSCOMPILING))
-mxnet_option(USE_OPERATOR_TUNING  "Enable auto-tuning of operators" ON IF NOT MSVC)
-mxnet_option(USE_GPERFTOOLS       "Build with GPerfTools support" OFF)
-mxnet_option(USE_JEMALLOC         "Build with Jemalloc support"   ON)
-mxnet_option(USE_DIST_KVSTORE     "Build with DIST_KVSTORE support" OFF)
-mxnet_option(USE_PLUGINS_WARPCTC  "Use WARPCTC Plugins" OFF)
-mxnet_option(USE_PLUGIN_CAFFE     "Use Caffe Plugin" OFF)
-mxnet_option(USE_CPP_PACKAGE      "Build C++ Package" OFF)
-mxnet_option(USE_MXNET_LIB_NAMING "Use MXNet library naming conventions." ON)
-mxnet_option(USE_GPROF            "Compile with gprof (profiling) flag" OFF)
-mxnet_option(USE_CXX14_IF_AVAILABLE "Build with C++14 if the compiler supports it" OFF)
-mxnet_option(USE_VTUNE            "Enable use of Intel Amplifier XE (VTune)" OFF) # one could set VTUNE_ROOT for search path
-mxnet_option(USE_TVM_OP           "Enable use of TVM operator build system." OFF)
-mxnet_option(ENABLE_CUDA_RTC      "Build with CUDA runtime compilation support" ON)
-mxnet_option(BUILD_CPP_EXAMPLES   "Build cpp examples" ON)
-mxnet_option(INSTALL_EXAMPLES     "Install the example source files." OFF)
-mxnet_option(USE_SIGNAL_HANDLER   "Print stack traces on segfaults." ON)
-mxnet_option(USE_TENSORRT         "Enable inference optimization with TensorRT." OFF)
-mxnet_option(USE_ASAN             "Enable Clang/GCC ASAN sanitizers." OFF)
-mxnet_option(ENABLE_TESTCOVERAGE  "Enable compilation with test coverage metric output" OFF)
-mxnet_option(USE_INT64_TENSOR_SIZE "Use int64_t to represent the total number of elements in a tensor" OFF)
-mxnet_option(BUILD_CYTHON_MODULES "Build cython modules." OFF)
+option(USE_CUDA "Build with CUDA support"   ON)
+option(USE_OLDCMAKECUDA "Build with old cmake cuda" OFF)
+option(USE_NCCL "Use NVidia NCCL with CUDA" OFF)
+option(USE_OPENCV "Build with OpenCV support" ON)
+option(USE_OPENMP "Build with Openmp support" ON)
+cmake_dependent_option(USE_CUDNN "Build with cudnn support" ON "USE_CUDA" OFF) # one could set CUDNN_ROOT for search path
+cmake_dependent_option(USE_SSE "Build with x86 SSE instruction support" ON "NOT ARM" OFF)
+option(USE_F16C "Build with x86 F16C instruction support" ON) # autodetects support if ON
+option(USE_LAPACK "Build with lapack support" ON)
+option(USE_MKL_IF_AVAILABLE "Use MKL if found" ON)
+cmake_dependent_option(USE_MKLDNN "Build with MKL-DNN support" ON
+  "USE_MKL_IF_AVAILABLE;NOT APPLE;NOT MSVC;CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL \"x86_64\";NOT CMAKE_CROSSCOMPILING" OFF)
```

Comment:
this line.cmake_dependent_option hidden USE_MKLDNN  on MSVC.but we needed is off by default.

Reply:
What do you mean? It is off by default, only if all of the following conditions are true, we set it to ON. That's exactly the same logic as we had before:

`"USE_MKL_IF_AVAILABLE;NOT APPLE;NOT MSVC;CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL \"x86_64\";NOT CMAKE_CROSSCOMPILING"`

vs before

`USE_MKL_IF_AVAILABLE AND (NOT APPLE) AND (NOT MSVC) AND (CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "x86_64") AND (NOT CMAKE_CROSSCOMPILING))`

The `;` stands for `AND`

Reply:
From the documentation of cmake_dependent_option, if the condition is false. The option will be hidden, which will cause MKLDNN to not be opened in the MSVC environment.

Reply:
Maybe I understand the document wrong. I'll do a test

Reply:
![image](https://user-images.githubusercontent.com/771744/70506062-6f86b480-1b64-11ea-8007-00b06f11f0e3.png)
not wrong.cmake_dependent_option hidden option with MSVC.

Reply:
Which document do you refer to? You can check the official documentation https://cmake.org/cmake/help/v3.16/module/CMakeDependentOption.html

> This macro presents an option to the user only if a set of other conditions are true. When the option is not presented a default value is used, but any value set by the user is preserved for when the option is presented again.

Thus the default value is used if the condition is false. Thus we have the same logic.

Reply:
I'm not familiar with the Windows UI. What does your screenshot show?

Reply:
Do you mean we want to allow users on MSVC to use MKLDNN? Then indeed the logic is wrong.

Reply:
The operation logic of mxnet_optiond has been modified. The new version of mxnet_option does not hide the options. It just adjusts the default values ​​based on the conditions.
https://github.com/apache/incubator-mxnet/pull/14743

Reply:
If it is supported, why does it have a different default value? Let's use the same default value (`ON`) then on all supported platforms? If it is not supported, then hiding is correct.

Reply:
Let's discuss the assimilation of default values separately. I fixed the hiding problem for now with `if else` condition. So now it's at least very easy to understand what is going on (whereas before users need to study the Macro implementation to understand the behaviour..)